### ISSUE TITLE
Flacky test because of race condition

### ISSUE THREAD

**Issue by MikaelMayer** [dafny-lang/dafny#3502]
I found another race condition that led to the LSP test `EnsuresManyDocumentsCanBeVerifiedAtOnce ` failing, which could happen on any other tests, it's just that this one is the most stressful for the system.

Here is the diff between expected and actual
```
  .  |  |  |  I  |  | :predicate F(i: int) {
  .  |  |  |  I  |  | :  false // Should not be highlighted in gutter.
  .  |  |  |  I  |  | :}
     |  |  |  I  |  | :
- .  S [S][ ][I][S][ ]:method H()
- .  S [=][=][-][~][O]:  ensures F(1)
- .  S [=][=][-][~][=]:{//Next: { assert false;
- .  S [S][ ][I][S][ ]:}
+ . [I][S][ ][I][S][ ]:method H()
+ . [=][~][=][-][~][O]:  ensures F(1)
+ . [=][~][=][-][~][=]:{//Next: { assert false;
+ . [I][S][ ][I][S][ ]:}
```
Looking at it, it means that the status of the method was set to "running" _after_ the first assertion batch was verified. I can witness I have seen this behavior in real instances of VSCode, where it was saying it verified, and then it pretended it started verifying only to provide the same result afterwards.

This comes from line 258 of Compilation.cs:
https://github.com/dafny-lang/dafny/blob/f224f9a15f2a225883570eea8a20f1b69bd74dbd/Source/DafnyLanguageServer/Workspace/Compilation.cs#L253-L260

Basically, it's a place where we tell the gutter icons that the implementation has started, whereas the boogie engine had already started, send the notification that the implementation had started but it was ignored, and sent the indication about the result of verifying the assertion batch, which was taken into account. Here is the code that indicates that the original implementation verification started message was ignored.

https://github.com/dafny-lang/dafny/blob/f224f9a15f2a225883570eea8a20f1b69bd74dbd/Source/DafnyLanguageServer/Language/AssertionBatchCompletedObserver.cs#L51-L52

Originally (back in July), this code was the only ground truth to tell if an implementation had started verifying.
